fpdf/table.py

```diff
@@ -0,0 +1,337 @@
+        if lines_heights_only:
+            return lines_heights
+
+    def _get_col_width(self, i, j, colspan=1):
+        if not self._col_widths:
+            cols_count = self._rows[i].cols_count
+            return colspan * (self._width / cols_count)
+        if isinstance(self._col_widths, Number):
+            return colspan * self._col_widths
+        if j >= len(self._col_widths):
+            raise ValueError(
+                f"Invalid .col_widths specified: missing width for table() column {j + 1} on row {i + 1}"
+            )
+        # pylint: disable=unsubscriptable-object
+        col_width = 0
+        for k in range(j, j + colspan):
+            col_ratio = self._col_widths[k] / sum(self._col_widths)
+            col_width += col_ratio * self._width
+        return col_width
+
+    def _get_lines_heights_per_cell(self, i) -> List[List[int]]:
+        row = self._rows[i]
+        lines_heights = []
+        for j in range(len(row.cells)):
+            lines_heights.append(
+                self._render_table_cell(
+                    i,
+                    j,
+                    cell_line_height=self._line_height,
+                    row_height=self._line_height,
+                    lines_heights_only=True,
+                )
+            )
+        return lines_heights
+
+
+class Row:
+    "Object that `Table.row()` yields, used to build a row in a table"
+
+    def __init__(self):
+        self.cells = []
+
+    @property
+    def cols_count(self):
+        return sum(cell.colspan for cell in self.cells)
+
+    def cell(
+        self, text="", align=None, style=None, img=None, img_fill_width=False, colspan=1
+    ):
+        """
+        Adds a cell to the row.
+
+        Args:
+            text (str): string content, can contain several lines.
+                In that case, the row height will grow proportionally.
+            align (str, fpdf.enums.Align): optional text alignment.
+            style (fpdf.fonts.FontStyle): optional text style.
+            img: optional. Either a string representing a file path to an image,
+                an URL to an image, an io.BytesIO, or a instance of `PIL.Image.Image`.
+            img_fill_width (bool): optional, defaults to False. Indicates to render the image
+                using the full width of the current table column.
+            colspan (int): optional number of columns this cell should span.
+        """
+        if text and img:
+            raise NotImplementedError(
+                "fpdf2 currently does not support inserting text with an image in the same table cell."
+                "Pull Requests are welcome to implement this 😊"
+            )
+        cell = Cell(text, align, style, img, img_fill_width, colspan)
+        self.cells.append(cell)
+        return cell
+
+
+@dataclass
+class Cell:
+    "Internal representation of a table cell"
+    text: str
+    align: Union[str, Align]
+    style: FontStyle
```

Comment:
The `style=` argument stands in the way of combining text with different fonts later. This information should be taken from the global **current font** (and modified at the rendering stage where necessary eg. for headers).

Reply:
We currently need this for `HTML2PDF`: https://github.com/PyFPDF/fpdf2/blob/table/fpdf/html.py#L284

In your opinion, how should we implement this now to best be future-proof?

Reply:
The solution is to set the font/style before calling `row.cell()` (or inside its context when using `cell.write()`).

`TextRegion.__enter__()` sets `pdf.page = 0`, and resets it on `__exit__()` before starting to render, in order to prevent state changes from getting prematurely written to the PDF.
`Table()` will have to do the same. Then each `Cell()` just remembers the current style instead of accepting it as an argument. And in the end you proceed and render as currently planned.

`_render_table_cell()` will have to check if a row (eg. a header) wants to force a different style (eg. bold), and use that  instead in this case.

Reply:
> The solution is to set the font/style before calling `row.cell()` (or inside its context when using `cell.write()`).

This won't work, as the table rendering is only performed at the end of the `FPDF.table()` context-manager, when `Table.render()` is called.

This is why cells style needs to be provided in a manner that can be stored inside a `Table` instance,
so that it can be taken into account later on, during `render()` execution.

Reply:
Have you taken my full explanation into acount? Including the `pdf.page = 0` part?

```python
class Cell:
    def __init__(self, pdf, ...):
        self.style = pdf.font_style
        ...
```
I don't see a reason why this shouldn't work. The same approach works fine in `TextRegion()`.

Reply:
You are right, I did not read your last message carefully enough.
Alright, this should work indeed.
However, it may have a cost in terms of memory usage if we store, for each cell of the table,
a copy of the current font family + emphasis (bold/italics/underline style) + font size + font color + fill color (background).

In terms of user-API, it seemed handy to provide styling parameters as arguments to `Row.cell()`,
instead of relying on the more verbose `use_font_style()` context-manager,
or on direct changes to the `FPDF` attributes (`.font_family`, `.font_style`, `.font_size_pt`...).
If in the end we capture the local values of those settings and store them in an attribute of `Cell`,
wouldn't it make sense to provide access to this attribute as an argument of the `Cell` constructor.

Could you please provide details on why you stated this @gmischler:
> The `style=` argument stands in the way of combining text with different fonts later

Having multiple ways to configure things can sometimes become confusing for end users
(_"There should be one-- and preferably only one --obvious way to do it."_)
Maybe that's the case here, and we should strive for consistency with the existing methods to configure font style.
I'm not sure.

One last thing that come to my mind: the `FontStyle` class introduced in this PR in the `fonts` module
only covers 5 font settings (font family + emphasis + font size + font color + fill color).
Should we also capture other current settings inside a `Cell` instance?
I'm thinking about `.font_stretching`, `.char_spacing`, `.text_mode`...

Reply:
> However, it may have a cost in terms of memory usage if we store, for each cell of the table,
a copy of the current font family + emphasis (bold/italics/underline style) + font size + font color + fill color (background).

Actually, there's a simpler way: Don't store the original string, but immediately turn it into a `Fragment()`. That already includes all the relevant information from the state stack. This is a memory use that you have in any case.

> In terms of user-API, it seemed handy to provide styling parameters as arguments to Row.cell(),
instead of relying on the more verbose use_font_style() context-manager,

Remember that the goal is to allow text of different fonts/sizes/styles/etc. to be placed in a table cell. This *requires* the use of the existing attribute setting methods.
I understand your wish to have "shortcut" ways of doing simple things, but implementing those first seems to be problematic. It is hard to recognize all the consequences and requirements of the full implementation before it actually exists. I recommend to implement the future-proof version first (in your eyes the clumsy one :wink:), and then to figure out which simplifications make sense and are reasonable to implement.

As to `FontStyle()` and `use_font_style()`: I don't quite see the point of those, to be honest. They seem completely redundant to other ways of reaching the same purpose. Adding overlapping functionality will not really make the library easier to use for the users, but is rather prone to confuse them.

Yes, as a general concept, I can see the appeal of having predefined text styles (as they typically exist in word processors).
But then, I don't remember ever seeing a user request for such a thing. fpdf2 is not a word processor for end users, we are offering functionality at a much lower level for developers.  And even more importantly, implementing such a high-level functionality as a special feature for table cells is probably not the right way to go about it. 

I recommend to implement the table functionality with the existing tools as far as possible, and to focus on making it work correctly and future-proof first. Once that is successful, if there are reasonable ways to add shortcuts, by all means add those as well. But starting the implementation with the shortcuts and trying to add the full functionality after that doesn't look like a good recipe for success.

If at a later stage there is a real demand for high-end text style management features, then we can try to figure out a concept for those that is easy to understand for users, and works well with all parts of the library, and not just table cells.

Reply:
> Actually, there's a simpler way: Don't store the original string, but immediately turn it into a Fragment(). That already includes all the relevant information from the state stack. This is a memory use that you have in any case.

As much as I like the idea of reusing existing code, already present in `fpdf2`, instead of using new abtractions,
I have trouble seing how we could use `Fragment` instances in a way that is "simpler"...

As a class, it is true that a `Fragment` already covers all of the properties contained in the `FontStyle` class I introduced.
However I see two problems rising when trying to get rid of `FontStyle` in favour of `Fragment`:

* `Fragment` cannot be used by the end user to define the table headings style. Easy table headings styling is something that I really want to provide to our users, so that they do not have to style by themselves the first line of cells in the table, for example by using some `if i == 0` special case in the loop that renders rows. Moreover, `Table` needs some knowledge of those headings (and their style) to be able to repeat them after a page break. How would you go on to implement this without introducing a `FontStyle` class @gmischler?

* if a `Cell` has a `Fragment` attribute, the consequence will be that it will have to be rendered with `FPDF._render_styled_text_line()`, which is the only method that currently takes `Fragment` as input. This means that `table.py` won't be able to benefit from a significant part of the code of `multi_cell`, that handles `markdown`, `split_only`, multi-line breaks... As a consequence, all this code would have to be duplicated in `table.py` in order to implement table cell rendering. That is certainly doable, but not ideal to avoid code duplication and for maintainability. In the end, it seems to me that this is about prefering to reuse the existing `multi_cell()` method or the existing `Fragment` class, and none of those options is really perfect, they both have drawbacks.

> I can see the appeal of having predefined text styles (as they typically exist in word processors).
> But then, I don't remember ever seeing a user request for such a thing

Agreed. That was never the ambition of `FontStyle`. It was just a means to an end, in the context of crafting the `table()` API.

However, regarding library design & composability, if we put aside the current behaviour of `FPDF` and its dozens of attributes, in insight it would probably have been better to gather font styling settings in a dedicated class, that could have then been passed to various rendering methods & classes, instead of storing them directy as properties of the main `FPDF` class. In that sense, a `FontSettings` / `FontStyle` / `FontProperties` class would make sense. But this subject is not the main goal of this PR!

> As to `FontStyle()` and `use_font_style()`: I don't quite see the point of those, to be honest. They seem completely redundant to other ways of reaching the same purpose. Adding overlapping functionality will not really make the library easier to use for the users, but is rather prone to confuse them.

I kind of agree on this. But as I mentioned above, I needed a way to provide the user some style for the table headings, and store this style inside an intance of `Table`. I took inspiration from the existing [TitleStyle](https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.TitleStyle) class to build the parent `FontStyle` class.

I'd be willing to get rid of both `FontStyle` & `use_font_style()` if we can figure another way to achieve this goal of tyling table headings.

---

I think we share the same vision regarding what's best for the library @gmischler: avoid duplication of functionalities, make it straightforward for the user to the use it, provide useful features and keep it all consistent.
I'd really like to know if you have other practical suggestions regarding table headings and how to provide styling for those.
The more I think about it, the more I get convinced that a `Fragment` instance won't do it, as we will need to dissociate the text string from the font properties that it contains.
Maybe we should refactor a `Fragment` this way:
````python
class Fragment:
    def __init__(self, characters, graphics_state, k, url):
        self.characters = characters
        self.font_style = FontStyle(graphics_state, k)
        self.url = url
```